### PR TITLE
[Translations] Fix permissions check and add userOwner & userModification columns

### LINKED ISSUE #12722: [Bug]: shared translations does not apply language permissions
### Expected behavior

When a user has view permissions for a language he should not be able to edit the shared translations for that language.

### Actual behavior

The user is able to edit the restricted language.

### Steps to reproduce

Tested on Pimcore 10.5. Unable to test this on the demo environment as it requires access to the users/roles.

Create a new user
- Permissions: enable 'Translations'
- Shared Translation Settings > Language Permissions: enable view access for one or more languages
- Save

Language_permissions

- Log in as the new user
- Go to shared translations
- Edit a translation for a restricted language

Shared_translations_edit

The same happens when setting the language permissions on a role and assigning the role to a user.

In Pimcore 6 the bug already existed partially. Directly editing the cell by double clicking was blocked, but you were still able to right click and edit the text/HTML.

### PR DESCRIPTION
**Pull request by dvesh3**
## Changes in this pull request  
Resolves #12722

### CODE REVIEW

File: models/Translation/Dao.php [hunk 1]
```diff
@@ -173,8 +185,32 @@ public function createOrUpdateTable()
...
+        $updateTime = time();
+        $this->model->setModificationDate($updateTime);
+
+        if (!$this->model->getCreationDate()) {
+            $this->model->setCreationDate($updateTime);
+        }
+
+        // auto assign user if possible, if no user present, use ID=0 which represents the "system" user
+        $userId = 0;
+        $user = \Pimcore\Tool\Admin::getCurrentUser();
```
Comment (kjkooistra-youwe) [inline models/Translation/Dao.php]: Alternatively you could maybe do something like:
`$userId = \Pimcore\Tool\Admin::getCurrentUser()?->getId() ?? 0`

**Review by kjkooistra-youwe — APPROVED** [review]

**Comment by kjkooistra-youwe** [comment]
Patched local environment. Langues are now blocked correctly.
Translations

File: models/Translation/Dao.php (hunk 1 above)
Comment (dvesh3) [inline models/Translation/Dao.php]:
```suggestion
        $userId = \Pimcore\Tool\Admin::getCurrentUser()?->getId() ?? 0;
```

File: models/Translation/Dao.php [hunk 2]
```diff
@@ -173,8 +185,31 @@ public function createOrUpdateTable()
...
+
+        if (!$this->model->getCreationDate()) {
+            $this->model->setCreationDate($updateTime);
+        }
+
+        // auto assign user if possible, if no user present, use ID=0 which represents the "system" user
+        $userId = \Pimcore\Tool\Admin::getCurrentUser()?->getId() ?? 0;
+        if ($user instanceof Model\User) {
+            $userId = $user->getId();
+        }
```
Comment (dvesh3) [inline models/Translation/Dao.php]:
```suggestion
```

**Comment by dvesh3** [comment]
> Patched local environment. Langues are now blocked correctly. Translations

@kjkooistra-youwe thanks for the confirmation! this will be part of next bugfix release 10.5.1.

File: models/Translation.php [hunk 3]
```diff
@@ -68,6 +68,21 @@ final class Translation extends AbstractModel
...
+     * @var int|null
+     */
+    protected ?int $userOwner = null;
+
+    /**
+     * ID of the user who make the latest changes
+     *
+     * @var int|null
+     */
+    protected $userModification;
```
Comment (blankse) [inline models/Translation.php]:
```suggestion
    protected ?int $userModification = null;
```

File: models/Translation.php [hunk 4]
```diff
@@ -200,6 +215,38 @@ public function setDomain(string $domain): void
...
+     */
+    public function getUserOwner(): ?int
+    {
+        return $this->userOwner;
+    }
+
+    /**
+     * @param int|null $userOwner
+     */
+    public function setUserOwner(?int $userOwner)
```
Comment (blankse) [inline models/Translation.php]:
```suggestion
    public function setUserOwner(?int $userOwner): void
```

File: models/Translation.php [hunk 5]
```diff
@@ -200,6 +215,38 @@ public function setDomain(string $domain): void
...
+    {
+        $this->userOwner = $userOwner;
+    }
+
+    /**
+     * @return int|null
+     */
+    public function getUserModification()
```
Comment (blankse) [inline models/Translation.php]:
```suggestion
    public function getUserModification(): ?int
```

File: models/Translation.php [hunk 6]
```diff
@@ -200,6 +215,38 @@ public function setDomain(string $domain): void
...
+    {
+        return $this->userModification;
+    }
+
+    /**
+     * @param int $userModification
+     */
+    public function setUserModification($userModification)
```
Comment (blankse) [inline models/Translation.php]:
```suggestion
    public function setUserModification(?int $userModification)
```

File: models/Translation/Dao.php [hunk 7]
```diff
@@ -173,8 +185,28 @@ public function createOrUpdateTable()
...
                           `creationDate` int(11) unsigned DEFAULT NULL,
                           `modificationDate` int(11) unsigned DEFAULT NULL,
+                          `userOwner` int(11) unsigned DEFAULT NULL,
+                          `userModification` int(11) unsigned DEFAULT NULL,
                           PRIMARY KEY (`key`,`language`),
                           KEY `language` (`language`)
                         ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;");
     }
+
+    protected function updateModificationInfos()
```
Comment (blankse) [inline models/Translation/Dao.php]:
```suggestion
    protected function updateModificationInfos(): void
```

File: models/Translation.php [hunk 8]
```diff
@@ -200,6 +215,38 @@ public function setDomain(string $domain): void
...
+    /**
+     * @return int|null
+     */
+    public function getUserModification()
+    {
+        return $this->userModification;
+    }
+
+    /**
+     * @param int $userModification
```
Comment (blankse) [inline models/Translation.php]:
```suggestion
     * @param int|null $userModification
```

File: models/Translation.php (hunk 4 above)
Comment (dvesh3) [reply models/Translation.php]: we don't add `void` return type-hints for simplicity.

File: models/Translation.php (hunk 6 above)
Comment (dvesh3) [reply models/Translation.php]: please revert `void`

File: models/Translation/Dao.php (hunk 7 above)
Comment (dvesh3) [reply models/Translation/Dao.php]: we don't add void return type-hints for simplicity.

File: models/Translation.php (hunk 6 above)
Comment (blankse) [reply models/Translation.php]: I changed it

File: models/Translation.php (hunk 4 above)
Comment (jdreesen) [reply models/Translation.php]: What? :confused:

Comment (dvesh3) [reply models/Translation.php]: just to be consistent with other models as we avoid the overhead of adding `void` return type to the existing methods in models.

Comment (blankse) [reply models/Translation.php]: @dvesh3 Than we can't get a higher PhpStan level. The next level 6 want to set return types. When there is no `: void` you have to set `@return void` in the php doc

Comment (dvesh3) [reply models/Translation.php]: Ok, that's a good point. @fashxp @brusch I think we to consider this change for our models. wdyt?

Comment (brusch) [reply models/Translation.php]: @dvesh3 yep, that's a valid point, better having proper ` : void` than `@return void` imho.

File: models/Translation.php [hunk 9]
```diff
@@ -200,6 +215,38 @@ public function setDomain(string $domain): void
...
+     */
+    public function getUserModification(): ?int
+    {
+        return $this->userModification;
+    }
+
+    /**
+     * @param int|null $userModification
+     */
+    public function setUserModification(?int $userModification)
```
Comment (blankse) [inline models/Translation.php]:
```suggestion
    public function setUserModification(?int $userModification): void
```

File: models/Translation.php (hunk 4 above)
Comment (dvesh3) [reply models/Translation.php]: We decided to add `:void` return typehints in new methods and then we can add `@return void` phpdoc to models on Pimcore 11 to upgrade phpstan level 6.

**Review by aashan10 — APPROVED** [review]

**Pull request merged**